pennylane/ops/op_math/prod.py

```diff
@@ -247,30 +242,41 @@ def decomposition(self):
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
 
-        def mats_gen():
-            for ops in self.overlapping_ops:
-                if len(ops) == 1:
-                    yield (
-                        qml.matrix(ops[0])
-                        if isinstance(ops[0], qml.Hamiltonian)
-                        else ops[0].matrix()
-                    )
+        mats = []
+        for ops in self.overlapping_ops:
+            if len(ops) == 1:
+                batched = ops[0].batch_size is not None
+                mat = qml.matrix(ops[0]) if isinstance(ops[0], qml.Hamiltonian) else ops[0].matrix()
```

Comment:
Just a comment: I think this logic `qml.matrix(H) if isinstance(H, qml.Hamiltonian) else H.matrix()` exists a lot in the codebase. Maybe we want to create a functionality for this at some point that doesn't require writing this out every time?

Reply:
I think `qml.matrix` _is_ that functionality, but I'm guessing people avoid just doing `qml.matrix` because of the overhead of it being wrapped as an op_transform?

Reply:
This I think may also be solved by the new op arithmetic, in that `qml.matrix()` will no longer be 'opinionated', but will simply always try and compute the matrix.

Which... has other downfalls. The matrix computation could take forever if you are not careful.

Reply:
this makes me wonder - I have more than once in the past taken advantage of the fact that `qml.matrix()` works with tapes (because of the `tape_transform` side effect defined as `_matrix()`). Perhaps that functionality should be explicitly separated, and the user should _know_ they're looking for the matrix of a tape/series of operations. With that separated out, we would simply use `op.matrix()` (or `tape.matrix()`) when relevant.

Reply:
Honestly I don't know why we are doing this...

I just profiled both functions and we don't have that much overhead:
```pycon
>>> import pennylane as qml
>>> import timeit
>>> prod = qml.prod(qml.PauliX(0), qml.PauliY(1))
>>> %timeit prod.matrix()
22.1 µs ± 90.3 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
>>> %timeit qml.matrix(prod)
24.4 µs ± 227 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

I changed this line to use `qml.matrix`.

Reply:
I feel like @albi3ro could have an additional insight and opinion here as well, having thought a lot about operator arithmetic.

Reply:
So Hamiltonian is a special case where it just never got `Hamiltonian.matrix` defined. I think this idea was "this could potentially be expensive, so we don't want to define it on the operation.". I do have an issue open to move `qml.utils.sparse_hamiltonian` to the `Hamiltonian.sparse_matrix` method. This has just been a low-priority code organization improvement.

I think we should use the `Operator.matrix` method internally, especially in cases like this. 

If we were planning on maintaining `Hamiltonian` in the long term, I would say we should add `Hamiltonian.matrix`, but I don't think it's necessarily worth the effort now.

Reply:
Mmm ok. Maybe this `2us` difference can get noticeable when having tons of operands...

I'll revert it to using `op.matrix()`

Reply:
Should this conversation be closed? :)